flow/layers/shader_mask_layer.cc

```diff
@@ -33,16 +36,32 @@ void ShaderMaskLayer::Preroll(PrerollContext* context, const SkMatrix& matrix) {
   Layer::AutoPrerollSaveLayerState save =
       Layer::AutoPrerollSaveLayerState::Create(context);
   ContainerLayer::Preroll(context, matrix);
+
+  if (render_count_ >= kMinimumRendersBeforeCachingFilterLayer) {
+    TryToPrepareRasterCache(context, this, matrix);
+  } else {
+    render_count_++;
+    TryToPrepareRasterCache(context, GetCacheableChild(), matrix);
+  }
 }
 
 void ShaderMaskLayer::Paint(PaintContext& context) const {
   TRACE_EVENT0("flutter", "ShaderMaskLayer::Paint");
   FML_DCHECK(needs_painting(context));
+  if (context.raster_cache &&
+      context.raster_cache->Draw(this, *context.leaf_nodes_canvas)) {
+    return;
+  }
 
   AutoCachePaint cache_paint(context);
   Layer::AutoSaveLayer save = Layer::AutoSaveLayer::Create(
       context, paint_bounds(), cache_paint.paint());
-  PaintChildren(context);
+  if (!context.raster_cache ||
+      !context.raster_cache->Draw(GetCacheableChild(),
+                                  *context.leaf_nodes_canvas,
+                                  cache_paint.paint())) {
```

Comment:
This will apply the modifications to the paint twice. The AutoCachePaint sets up the paint for inherited opacity, but that only applies to the saveLayer, not the children. Just let the paint parameter default on this call.

Reply:
> 
Do you mean we use engine unittest to demonstrates that?